### ISSUE TITLE
`ObjectStore::get_range` returns `Bytes` instead of `GetResult`

### ISSUE THREAD

**Issue by Dandandan** [apache/arrow-rs#3632]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
While using the object store (S3 indirectly via DataFusion/Ballista) and using `get_range` requests, we can not stream the results but are required to fetch entire column chunks instead of being able to stream the data directly.

There are some limitations by not providing such a method in the `ObjectStore`
* Reducing memory usage/scalability by fetching / incrementally
* Processing more incrementally instead of waiting for the entire range to be fetched
* Being able to push down limits to every part of the query (e.g. row filters) and aborting early.

**Describe the solution you'd like**
Change the `get_range` implementation to return `GetResult` or introduce a new `get_range_stream` implementation, returning a stream, allowing to process / stream incrementally.

I checked this to be possible for S3, and I expect it to be for other implementations as well:
```rust
    async fn get_range_stream(&self, location: &Path, range: Range<usize>) -> Result<GetResult> {
        let stream = self
            .client
            .get_request(location, Some(range), false)
            .await?
            .bytes_stream()
            .map_err(|source| crate::Error::Generic {
                store: "S3",
                source: Box::new(source),
            })
            .boxed();
        Ok(GetResult::Stream(stream))
    }
```

We can provide a default implementation based on `get_range` in the meantime for other object stores.

**Describe alternatives you've considered**
-->

**Additional context**

**Comment by tustvold** [comment]
I don't think we can return `GetResult` as one of the variants is a `File` where a range would have unclear meaning. That being said from a technical standpoint I don't see anything wrong with a `get_range_stream`, however, I would highlight some motivations for the current design.

* Round-trip latency to object storage is largely agnostic to the response size, 1KB and 1MB will both take ~100ms. It is only once you get into >10MB territory where it starts to make an appreciable difference
* Parquet relies on `get_range**s**_ to coalesce multiple small `get_range` into a larger single request for performance and cost reasons, this is not possible with streams (or at the very least highly non-trivial)
* Related to the above, most formats that can perform pushdown are "chunked" that is the reader knows ahead of time what it needs to read, it is almost always more performant to read in units of this size and feed them into a synchronous decoder, than to interleave IO and decode

> push down limits to every part of the query (e.g. row filters) and aborting early.

I'm curious as to the data format you are intending for this, CSV and JSON support limit natively using the current streaming decode based around `get()`, it is unclear to me how `get_range` could be used with them.

Parquet supports arbitrary limit pushdown through its `RowSelection` API

**Comment by Dandandan** [comment]
Hi @tustvold, thanks for the response!

This would be for using the Parquet format.

We (Coralogix) were seeing some slower response times from queries with limits added, triggering this investigation. No `RowSelection` was present. That was added at our side (or a `RowFilter`), that seems for a large part to "fix" this issue indeed.

Data is in the ~1GB per file regime. I believe I often see a bit better latency than 100ms btw on S3 - can be as low as 30-40ms for smaller GET requests.

Some directions I still see some benefits of exposing a streaming result rather than collecting to bytes before being able to process.

* Reading very large files (without limits). Currently this uses more memory than necessary if data would be processed more incrementally. 
* Apply limits for row predicate evaluation as well. Currently it still requires fetching large parts of the data, while it might be able to only use a small subset of the data necessary to evaluate it.
* Concurrent IO / decode with IO prefetching -> currently we need to wait on the data to be available, which limits the amount of throughput / increases the latency to get the first results.

**Comment by tustvold** [comment]
> Currently this uses more memory than necessary if data would be processed more incrementally

It fetches row groups at a time, there isn't a smaller horizontal unit of IO that can be used, and nor is it actually a good idea. Small fetch requests destroy performance. This is the major thing that differentiates DataBrick's closed-source S3 parquet reader from the open source arrow one, and gives it significantly better performance.

> Apply limits for row predicate evaluation as well.

RowSelection provided this and more already

> Concurrent IO / decode with IO prefetching

The parquet reader would not be able to make use of this stream abstraction to achieve this as it relies on reading contiguous byte arrays and slicing them up for decode. Concurrent IO is definitely possible, however, an earlier version did this, it was mainly removed as it added additional memory pressure.

**Comment by tustvold** [comment]
Closing in favour of - https://github.com/apache/arrow-rs/issues/3632

**Comment by alamb** [comment]
Migrating from arrow-rs issue [#3632](https://github.com/apache/arrow-rs/issues/3632)
